pkg/blockstorage/vmware/vmware.go

```diff
@@ -465,6 +475,78 @@ func (p *FcdProvider) GetCategoryID(ctx context.Context, categoryName string) (s
 	return cat.ID, nil
 }
 
+// SnapshotsList is part of blockstorage.Provider
+func (p *FcdProvider) SnapshotsList(ctx context.Context, tags map[string]string) ([]*blockstorage.Snapshot, error) {
+	if filterStr := generateSnapshotDescription(tags); filterStr != "" {
+		volumeIDs := strings.Split(tags[VolumeIdListTag], ",")
+		if len(volumeIDs) == 0 {
+			return nil, errors.New("vSphere can't list by description without list of volumes. Cannot list snapshots")
+		}
+		return p.snapshotsListByDescription(ctx, volumeIDs, filterStr)
+	}
+	return p.snapshotsListByTag(ctx, tags)
+}
+
+func (p *FcdProvider) createSnapshotAndWaitForCompletion(volume blockstorage.Volume, ctx context.Context, description string) (types.AnyType, error) {
+	log.Debug().Print("CreateSnapshot", field.M{"VolumeID": volume.ID})
+	task, err := p.Gom.CreateSnapshot(ctx, vimID(volume.ID), description)
+	if err != nil {
+		return nil, errors.Wrap(err, "CreateSnapshot task creation failure")
+	}
+	log.Debug().Print("Started CreateSnapshot task", field.M{"VolumeID": volume.ID})
+	res, err := task.Wait(ctx, vmWareTimeout)
+	return res, err
+}
+
+func (p *FcdProvider) getCreatedSnapshot(ctx context.Context, description string, volID string, notEarlierThan time.Time) interface{} {
+	var filteredSns []blockstorage.Snapshot
+	sns, err := p.snapshotsListByDescription(ctx, []string{volID}, description)
+	for _, sn := range sns {
+		if notEarlierThan.Before((time.Time)(sn.CreationTime)) {
+			filteredSns = append(filteredSns, *sn)
+		}
+	}
+
+	if len(filteredSns) == 1 && err == nil {
+		return filteredSns[0]
+	}
+	if err != nil {
+		log.Error().WithError(err).Print("Failed to list when checking failed creation")
+	}
+	if len(filteredSns) > 1 {
+		log.Error().Print(fmt.Sprintf("More than one snapshot was found, IDs: %s", strings.Join(getSnapshotsIDs(filteredSns), ",")))
+	}
+	return nil
+}
+
+func generateSnapshotDescription(tags map[string]string) string {
+	var tagsAsStr []string
+	for name, value := range tags {
+		if strings.HasPrefix(name, DescriptionTag) {
+			tagsAsStr = append(tagsAsStr, fmt.Sprintf("%s:%s", name, value))
+		}
+	}
+	return strings.Join(tagsAsStr, ",")
+}
+
+func getNotDscriptionTags(tags map[string]string) map[string]string {
+	result := make(map[string]string)
+	for name, value := range tags {
+		if !strings.HasPrefix(name, DescriptionTag) {
+			result[name] = value
+		}
+	}
+	return result
+}
+
+func getSnapshotsIDs(snapshots []blockstorage.Snapshot) []string {
+	var result []string
```

Comment:
For the same reasoning as above.

```suggestion
	result := make([]string, 0, len(snapshots))
```